### ISSUE TITLE
output of cost (loss) for validation

### ISSUE THREAD

**Issue by zhaogengyan** [Kamnitsask/deepmedic#102]
Hi Konstantinos,

Thank you for sharing the work of Deepmedic, again!
My question is: I don't know why in the output log for training there is no cost (loss) for the validation dataset. There are cost and accuracy for the training data, but only accuracy for the validation data.

In the file deepmedic/deepmedic/routines/training.py, it is said that
"In the case of VALIDATION, meanCostOfSubepoch is just a placeholder. Only valid when training."

It's pretty common to plot the loss of the training set and validation set together to check overfitting, right?

Thank you in advance.

Best,
Gengyan

**Comment by Kamnitsask** [comment]
Hello,

You are right. I have always been using the "accuracy" metrics and the difference in behaviour between train/test accuracies to observe convergence and possible overfitting, and not the loss, since accuracy is more intuitive to report. Note that cross entropy and accuracy over the same samples is very straightforwardly related.

If for some reason you badly need the loss from validation, you can quite easily log it and plot it similar to the training loss. Have a look at the file:
deepmedic->neuralnet->cnn3d.py
and the functions compileTrainFunction() and compileValidationFunction(). You ll have to change the latter to return the "cost" similar to the former.
Then you ll need to change deepmedic->routines->training.py -> doTrainOrValidationOnBatchesAndReturnMeanAccuraciesOfSubepoch()
which calls the above two functions and logs loss.

Hope the above helps.
K

**Comment by stalhabukhari** [comment]
Hi @Kamnitsask,

I observed that the code structure has changed significantly since this discussion. I also wish to log **per-epoch mean validation loss**. For this, I made the following changes:

`neuralnet/cnn3d.py > Cnn3d.setup_ops_n_feeds_to_val()`:
- `self._feeds_main['val']['cost'] = total_cost` where `total_cost` is received as an additional argument to the function.

`frontEnd/trainSession.py > TrainSession.run_session()`:
- Passed `trainer.get_total_cost()` to `cnn3d.setup_ops_n_feeds_to_val()`.

`routines/training.py > process_in_batches()`:
- `list_of_ops = [ops_to_fetch['cost']] + ops_to_fetch['list_rp_rn_tp_tn']` in the validation portion.
- Appropriately edited `results_of_run`, `cost_this_batch`, and `list_RpRnPpPn_per_class` in the validation portion.
- `mean_cost_subep = np.mean(costs_of_batches)`, to work for validation loss as well.

I now get the following error:
```
InvalidArgumentError (see above for traceback): You must feed a value for placeholder tensor 'net/inp_x_sub_0_train' with dtype float and shape [?,?,?,?,?]
	 [[node net/inp_x_sub_0_train (defined at /media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/neuralnet/cnn3d.py:254) ]]
	 [[node Sum_33 (defined at /media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/neuralnet/layers.py:505) ]]
```

I should state here that the code worked perfectly before making these changes. Kindly guide about what I have missed so far.

Regards,
Talha

**------------------------------------------------------------------------------------------------------------------------------------------**

**Complete error message:**
```
 ERROR: Caught exception in do_training(): You must feed a value for placeholder tensor 'net/inp_x_sub_0_train' with dtype float and shape [?,?,?,?,?]
	 [[node net/inp_x_sub_0_train (defined at /media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/neuralnet/cnn3d.py:254) ]]
	 [[node Sum_33 (defined at /media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/neuralnet/layers.py:505) ]]

Caused by op 'net/inp_x_sub_0_train', defined at:
  File "./deepMedicRun", line 178, in <module>
    session.run_session(sess_device, model_params, args.reset_trainer)
  File "/media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/frontEnd/trainSession.py", line 111, in run_session
    cnn3d.make_cnn_model(*model_params.get_args_for_arch())
  File "/media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/neuralnet/cnn3d.py", line 361, in make_cnn_model
    self._setupInputXTensors()
  File "/media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/neuralnet/cnn3d.py", line 254, in _setupInputXTensors
    self._inp_x['train']['x_sub_'+str(subpath_i)] = tf.placeholder(dtype="float32", shape=[None, None, None, None, None], name="inp_x_sub_"+str(subpath_i)+"_train")
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/ops/array_ops.py", line 2077, in placeholder
    return gen_array_ops.placeholder(dtype=dtype, shape=shape, name=name)
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/ops/gen_array_ops.py", line 5791, in placeholder
    "Placeholder", dtype=dtype, shape=shape, name=name)
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3300, in create_op
    op_def=op_def)
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1801, in __init__
    self._traceback = tf_stack.extract_stack()

InvalidArgumentError (see above for traceback): You must feed a value for placeholder tensor 'net/inp_x_sub_0_train' with dtype float and shape [?,?,?,?,?]
	 [[node net/inp_x_sub_0_train (defined at /media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/neuralnet/cnn3d.py:254) ]]
	 [[node Sum_33 (defined at /media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/neuralnet/layers.py:505) ]]


Traceback (most recent call last):
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1334, in _do_call
    return fn(*args)
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1319, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1407, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.InvalidArgumentError: You must feed a value for placeholder tensor 'net/inp_x_sub_0_train' with dtype float and shape [?,?,?,?,?]
	 [[{{node net/inp_x_sub_0_train}}]]
	 [[{{node Sum_33}}]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/routines/training.py", line 325, in do_training
    lbls_samples_per_path_val)
  File "/media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/routines/training.py", line 107, in process_in_batches
    results_of_run = sessionTf.run(fetches=list_of_ops, feed_dict=feeds_dict)
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 929, in run
    run_metadata_ptr)
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1152, in _run
    feed_dict_tensor, options, run_metadata)
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1328, in _do_run
    run_metadata)
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1348, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.InvalidArgumentError: You must feed a value for placeholder tensor 'net/inp_x_sub_0_train' with dtype float and shape [?,?,?,?,?]
	 [[node net/inp_x_sub_0_train (defined at /media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/neuralnet/cnn3d.py:254) ]]
	 [[node Sum_33 (defined at /media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/neuralnet/layers.py:505) ]]

Caused by op 'net/inp_x_sub_0_train', defined at:
  File "./deepMedicRun", line 178, in <module>
    session.run_session(sess_device, model_params, args.reset_trainer)
  File "/media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/frontEnd/trainSession.py", line 111, in run_session
    cnn3d.make_cnn_model(*model_params.get_args_for_arch())
  File "/media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/neuralnet/cnn3d.py", line 361, in make_cnn_model
    self._setupInputXTensors()
  File "/media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/neuralnet/cnn3d.py", line 254, in _setupInputXTensors
    self._inp_x['train']['x_sub_'+str(subpath_i)] = tf.placeholder(dtype="float32", shape=[None, None, None, None, None], name="inp_x_sub_"+str(subpath_i)+"_train")
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/ops/array_ops.py", line 2077, in placeholder
    return gen_array_ops.placeholder(dtype=dtype, shape=shape, name=name)
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/ops/gen_array_ops.py", line 5791, in placeholder
    "Placeholder", dtype=dtype, shape=shape, name=name)
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3300, in create_op
    op_def=op_def)
  File "/media/ee/DATA/Software/anaconda3/envs/tfDM/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1801, in __init__
    self._traceback = tf_stack.extract_stack()

InvalidArgumentError (see above for traceback): You must feed a value for placeholder tensor 'net/inp_x_sub_0_train' with dtype float and shape [?,?,?,?,?]
	 [[node net/inp_x_sub_0_train (defined at /media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/neuralnet/cnn3d.py:254) ]]
	 [[node Sum_33 (defined at /media/ee/DATA/Repositories/DeepMedic_Forked_17Feb20/deepmedic/neuralnet/layers.py:505) ]]
```

**Comment by stalhabukhari** [comment]
Hi,

I was able to output validation loss, and log it to TensorBoard.

This error resulted from calling the training data placeholders instead of the validation data placeholders during validation. For this, I defined another routine `get_total_cost_val()` in `trainer` class, which is the same as the previous routine but calls the validation data placeholders. Output from this routine is provided to `self._feeds_main['val']['cost']` as previously.

For TensorBoard logging, some simple changes in `logging/accuracyMonitor.py > __init__(), log_acc_subep_to_txt(), log_acc_subep_to_tensorboard(), report_metrics_samples_ep()` are needed.

Regards,
Talha
